### PR TITLE
[float8] Make module filter function for float8 training configurable

### LINKED ISSUE #1043: Make module filter function configurable for conversion to float8 training
This is currently [hardcoded](https://github.com/pytorch/torchtitan/blob/bbb1b616ef4d01b0ea745c13b03c192910a166c9/torchtitan/components/float8.py#L114) but we are thinking about supporting use cases where it will need to be configurable. For example, not converting linears in the routing layer of a MoE.

cc @tianyu-l

### PR DESCRIPTION
**Pull request by danielvegamyhre**
## Summary

Fixes #1043 

- Adds `--float8.filter_fqns` argument, which accepts a comma-separated list of FQNs. Defaults to empty list.
- When float8 training is enabled, any nn.Linear which has dim sizes divisible by 16 and whose FQN does not appear in the `filter_fqns` list, will be module-swapped for a Float8Linear layer.

## Test plan
- Manual testing the following cases:
    - Single FQN via command line: `--float8.filter_fqns "output"` - [logged module definition](https://www.internalfb.com/phabricator/paste/view/P1778127918)
    - Multiple FQNs via command line: `--float8.filter_fqns "output,attention.wq"` - [logged module defintion](https://www.internalfb.com/phabricator/paste/view/P1778128848)
    - Single FQN via toml ("output"): [logged module defintion](https://www.internalfb.com/phabricator/paste/view/P1778129382)
    - Multiple FQNs via toml ("output", "feed_forward.w1"): [logged module definition](https://www.internalfb.com/phabricator/paste/view/P1778134498)

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
cc @tianyu-l @drisspg

File: torchtitan/components/float8.py [hunk 1]
```diff
@@ -109,13 +110,26 @@ def convert_to_float8_training(self, model: nn.Module):
...
-            module_filter_fn=lambda mod, fqn: fqn != "output",
+            module_filter_fn=self._module_filter_fn,
         )
         logger.info(
             "Swapped to Float8Linear layers with enable_fsdp_float8_all_gather="
             f"{self.config.enable_fsdp_float8_all_gather}"
         )
 
+    def _module_filter_fn(self, mod, fqn):
+        assert isinstance(mod, nn.Linear)
```
Comment (drisspg) [inline torchtitan/components/float8.py]: We want to just return if it isn't a linear right?

Comment (danielvegamyhre) [reply torchtitan/components/float8.py]: Oh yeah it already should skip non-linears [here](https://github.com/pytorch/ao/blob/5802d2d2ca9b762cf61be1f8b6bf3c587cb1bc91/torchao/float8/float8_linear_utils.py#L41) so I was just asserting to act as an alert if that accidentally changes.

I think you're right though that returning False here makes more sense, so it'll just not be converted in that case. Updated it accordingly.

File: torchtitan/models/llama/train_configs/llama3_8b.toml [hunk 2]
```diff
@@ -60,3 +60,4 @@ selective_ac_option = 'op'  # 'int' = ac every positive int layer or 'op', ac ba
 [float8]
 enable_fsdp_float8_all_gather = false
 precompute_float8_dynamic_scale_for_fsdp = false
+target_fqns = ["output"]
```
Comment (tianyu-l) [inline torchtitan/models/llama/train_configs/llama3_8b.toml]:
```suggestion
filter_fqns = ["output"]
```

Comment (tianyu-l) [reply torchtitan/models/llama/train_configs/llama3_8b.toml]: could you also update this in all tomls?

Comment (danielvegamyhre) [reply torchtitan/models/llama/train_configs/llama3_8b.toml]: Done

**Comment by danielvegamyhre** [comment]
The CI test is failing due to the symmetric memory init issue I discovered and reported here: https://github.com/pytorch/pytorch/pull/149793#issuecomment-2763103641

It has been fixed in pytorch core, so I'd expect the torch nightly build used in CI to not have this issue. cc @tianyu-l @kwen2501 any idea what's going on here?
